Add LED Controller utility class

23adap opened the pull request:
add class for working with an led strip

23adap committed:
Create LEDController.java

add class for working with an led strip

23adap committed:
Update LEDController.java

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,64 @@
+import edu.wpi.first.wpilibj.AddressableLEDBuffer;
+
+/** 
+ * A class that encapsulates dealing with a strip of leds
```
Wording is a bit sketchy on this. Try to get rid of `A class that encapsulates`.

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,64 @@
+    private AddressableLEDBuffer m_ledBuffer;
+
+    //led PMW port
+    private int portIndex = 0;
```
This should be passed in the constructor, along with the buffer length

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,64 @@
+ */
+public class LEDController {
+
+    private AddressableLED m_led;
```
Use lowerCamelCase

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    // Sets the whole led strip to a single color.
+    public void setColor(int redIndex, int greenIndex, int blueIndex) {
```
In order to make the class easiest to use, methods should take in a [Color](https://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/util/Color.html).

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,64 @@
+    
+    private void setRainbow(int firstPixelHue, int saturation, int value) {
+        int m_rainbowFirstPixelHue = firstPixelHue;
+        for (var i = 0; i < m_ledBuffer.getLength(); i++) {
```
Try to avoid using `var` since it can be ambiguous

23adap committed:
Makes arguments for setColor more user-friendly and creates a blink method.

setColor will take in a color enum as defined by the java.awt.color class, which will then be converted into rgb values for the setRGB method. blink will cause the led stript to blink every second for a minute given a color as a parameter.

23adap committed:
Update LEDController.java

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,82 @@
+    private final AddressableLED mLED;
+    private final AddressableLEDBuffer mLEDBuffer;
+
+    public LEDController(final int portIndex, final int bufferLength) {
```
The `final` keyword doesn't do anything since the parameters are primitives (`int` parameters are already `final`)

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,82 @@
+    //creates a new thread for the execution of blinkColor
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+    // Sets the led strip to blink every second for a minute and takes the RBG values of the desired color as arguments
+    public void blinkColor(Color color) {
```
This method works, but it isn't very convenient to use (for example, you can't cancel the blink). Instead, you should write it in a way so that the LEDs are only updated when the method is called. When the method is called, it should check if enough time has elapsed to switch the leds on or off.

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,82 @@
+    }
+    
+    private void setRainbow(final int firstPixelHue, final int saturation, final int value) {
+        int m_rainbowFirstPixelHue = firstPixelHue;
```
This needs to be an instance variable (see the comment below)

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    // Sets the led strip to a rainbow by keeping the same saturation and value and cycling through the hue.
+    public void setRainbowPeriodic(final int firstPixelHue) {
```
You have the right idea here, but it's not implemented properly. Every time the setRainbowPeriodic method is called the rainbow should "move" by one unit. To achieve this you need to store the staring hue as an instance variable, rather than it being passed in. The example on the doc I sent should work.

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,82 @@
+    
+    private void setRainbow(final int firstPixelHue, final int saturation, final int value) {
+        int m_rainbowFirstPixelHue = firstPixelHue;
+        for (var i = 0; i < mLEDBuffer.getLength(); i++) {
```
var :(

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,82 @@
+        int m_rainbowFirstPixelHue = firstPixelHue;
+        for (var i = 0; i < mLEDBuffer.getLength(); i++) {
+            //calculate the hue
+            final var hue = (m_rainbowFirstPixelHue + (i * 180 / mLEDBuffer.getLength())) % 180;
```
var :(

23adap committed:
Fixes the rainbow method to move and creates an alternate blink option.

Creates two options for blinking: one to be called in RobotPeriodic, and one to be called on its own and to run until explicitly cancelled.

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,103 @@
+ */
+public class LEDController {
+
+    private final AddressableLED mLED;
```
Minor thing, don't start instance variable names with `m`, since we don't follow it in other places in the codebase

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,103 @@
+package harkerrobolib.subsystems;
```
Subsystems are usually mechanisms that play the robot game. The LEDController should be in the `util` package

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,103 @@
+    ScheduledFuture<?> runBlinker;
+    private long mLastTimeBlinkPeriodicSwitched;
+
+    private static final int BLINK_PERIOD_MILLIS = 1 * 1000;
```
This can be passed in as a parameter to the blink method

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,103 @@
+
+
+    //creates a new thread for the execution of blinkColor
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
```
Keep all the instance variables together

chiragzq replied:
Also maybe change the variable name to make it more clear what it schedules

chiragzq commented on src/harkerrobolib/subsystems/LEDController.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public void blinkPeriodic(Color color) {
+        if (System.currentTimeMillis() >= mLastTimeBlinkPeriodicSwitched + BLINK_PERIOD_MILLIS) {
```
Although this is correct, it would be more clear if it was structured like `currentTime - previousTime > period`

23adap committed:
Fixes stylistic issues.

chiragzq committed:
Add more documentation

ange1a-j14 commented:
looks good
